README.md
# Ether and Serial Communication Library

### Documentation

### Dependencies

### Install

### License

### Authors

### Contact


CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package comm
^^^^^^^^^^^^^^^^^^^^^^^^^^

0.0.1 (2023-03-31)
------------------
* setup


src/ether.cc
// this is for emacs file handling -*- mode: c++; indent-tabs-mode: nil -*-

// -- BEGIN LICENSE BLOCK -----------------------------------------------------------------------------------------------------

/*!
 *  Copyright (C) 2021, Autognity Srl, Via del Popolo 6, Orvieto (TR)
 *  \file ether.cpp
 *  \author Andrea Tamantini <andrea.tamantini@autognity.com>
 *  \date 2021-05-18
 */

// -- END LICENSE BLOCK -------------------------------------------------------------------------------------------------------


/*=============================================================================================================================
 * HEADER
 *===========================================================================================================================*/
#include <comm/ether.h>

namespace comm {

    Ether::Ether ( const string& address, uint16_t port, const string& eol, Timeout timeout ) : Comm(address,eol,timeout) {
        this->port_ = port;
        this->sockaddr_in_.sin_family = AF_INET;
        this->open();
    }

    // Read common function
    size_t Ether::read_ (uint8_t *data, size_t size) {
        // If the connection is not open and connected, throw
        if ( ! ( this->is_open_ && this->is_connected_ ) ) throw new ConnectionException ("Ether::read : not connected");
        // Pre-fill buffer with available bytes
        ssize_t bytes_read_now = ::recv ( this->fd_, data, size, MSG_DONTWAIT );
        size_t bytes_read = bytes_read_now > 0 ? bytes_read_now : 0;
        // Prepare timeout value : now + read + byte*size
        TimeCheck timeout ( this->timeout_.read, this->timeout_.byte, size );
        // Read until the desired size is read, there's nothing left to read or timeout expires
        while ( bytes_read < size ) {
            // If the timeout expired or i read no data in the last cycle break the reading loop
            if ( timeout.expired() || bytes_read_now == 0 ) break;
            // Wait for the device to be readable, otherwise check again on the next loop
            if ( this->waitRead_() < 1 ) continue;
            // Read new available bytes
            bytes_read_now = ::recv (this->fd_, data + bytes_read, size - bytes_read, MSG_DONTWAIT );
            // retry if interrupted
            if ( bytes_read_now == -1 && errno == EINTR) continue;
            // At least 1 byte should always be read
            if ( bytes_read_now < 1 )
                throw new InterfaceException (
                        "Ether::read : device reports readiness to read but returned no data, disconnected?", errno);
            bytes_read += bytes_read_now;
        }
        return bytes_read;
    }

    // Send common function
    size_t Ether::send_ (const uint8_t *data, size_t size) {
        // If the connection is not open and connected, throw
        if ( ! ( this->is_open_ && this->is_connected_ ) ) throw new ConnectionException ("Ether::send : not connected");
        // Prepare return variables
        ssize_t bytes_sent_now = ::send ( this->fd_, data, size, 0 );
        size_t bytes_sent = bytes_sent_now > 0 ? bytes_sent_now : 0;
        // Prepare timeout value : now + send + byte*size
        TimeCheck timeout ( this->timeout_.send, this->timeout_.byte, size );
        // Send until the desired size is sent or timeout expires
        while ( bytes_sent < size ) {
            // If the timeout expired break the reading loop
            if ( timeout.expired() ) break;
            // Wait for the device to be ready to receive, otherwise check again on the next loop
            if ( this->waitSend_() < 1 ) continue;
            // Send more byets
            bytes_sent_now = ::send (fd_, data + bytes_sent, size - bytes_sent, MSG_MORE );
            // retry if interrupted
            if ( bytes_sent_now == -1 && errno == EINTR) continue;
            // at least 1 byte should always be sent
            if (bytes_sent_now < 1)
                throw new InterfaceException (
                        "Ether::send : device reports readiness to receive but returned no data, disconnected?", errno);
            bytes_sent += bytes_sent_now;
        }
        return bytes_sent;
    }

    // Open the file descriptor
    void Ether::open_ () {
        if ( this->is_open_ ) 
            return;
        if ( ( this->fd_ = socket(this->sockaddr_in_.sin_family, SOCK_STREAM, IPPROTO_TCP) ) < 0 ) {
            switch ( errno ) {
            case EINTR: // Recurse because this is a recoverable error.
                this->open_ ( ); return;
            case ENFILE: case EMFILE: // Other error numbers
                throw new IOException ( "Ether::open : Too many file handles open", errno );
            default:
                throw new IOException ( "Ether::open : general IO exception", errno );
            }
        }
        // IS OPEN
        this->is_open_ = true;
    }
    // Establish a connection to the server
    void Ether::connect_ () {
        if ( this->is_connected_ || this->address_.empty() || this->port_ == 0 )
            return;
        // Set socket address
        set_address ( &(this->address_), this->port_, &(this->sockaddr_in_) );
        // UNLOCK
        set_options ( this->fd_, get_options(this->fd_) | O_NONBLOCK );
        // Trying to connect
        if ( ::connect( this->fd_, (struct sockaddr *) & ( this->sockaddr_in_ ), sizeof ( this->sockaddr_in_ ) ) < 0 )
            switch ( errno ) {
                case EINPROGRESS: {
                    fd_set fd_set_;
                    FD_ZERO ( &fd_set_ );
                    FD_SET ( this->fd_, &fd_set_ );
                    int result = select ( this->fd_+1, NULL, &fd_set_, NULL, & ( this->timeout_.conn ) );
                    if ( result < 1 )
                        throw new InterfaceException ( "Ether::connect : connection error", errno );
                } break;
                default:
                    throw new InterfaceException ( "Ether::connect : generic error", errno );
            }
        // LOCK
        set_options ( this->fd_, get_options(this->fd_) & ~O_NONBLOCK );
        // Set socket timeout
        this->setOptions_ ( );
        // IS CONNECTED
        this->is_connected_ = true;
    }
    // Set socket
    void Ether::setOptions_() {
        if ( setsockopt ( this->fd_, SOL_SOCKET, SO_RCVTIMEO,
                        (const char *) & ( this->timeout_.read ),
                        sizeof ( this->timeout_.read ) ) != 0 )
            throw new InterfaceException ( "Ether::setOptions : set read timeout", errno );
        if ( setsockopt ( this->fd_, SOL_SOCKET, SO_SNDTIMEO,
                        (const char *) & ( this->timeout_.send ),
                        sizeof ( this->timeout_.send ) ) != 0 )
            throw new InterfaceException ( "Ether::setOptions : set send timeout", errno );
        socklen_t socket_length = sizeof(int);
        int result;
        if ( getsockopt ( this->fd_, SOL_SOCKET, SO_ERROR, (void*)(&result), &socket_length ) < 0 )
            throw new InterfaceException ( "Ether::setOptions : select socket", errno );
        if ( result )
            throw new InterfaceException ( "Ether::setOptions : select socket", result );
    }

}

src/utils.cc
// this is for emacs file handling -*- mode: c++; indent-tabs-mode: nil -*-

// -- BEGIN LICENSE BLOCK -----------------------------------------------------------------------------------------------------

/*!
 *  Copyright (C) 2021, Autognity Srl, Via del Popolo 6, Orvieto (TR)
 *  \file ether.cpp
 *  \author Andrea Tamantini <andrea.tamantini@autognity.com>
 *  \date 2021-05-18
 */

// -- END LICENSE BLOCK -------------------------------------------------------------------------------------------------------


/*=============================================================================================================================
 * HEADER
 *===========================================================================================================================*/
#include <comm/utils.h>

namespace comm {

    timeval to_timeval ( double data ) {
        int64_t sec64 = static_cast<int64_t>(std::floor(data));
        if (sec64 < 0 || sec64 > std::numeric_limits<uint32_t>::max())
            throw std::runtime_error("Time is out of dual 32-bit range");
        struct timeval time{
            static_cast<uint32_t>(sec64),
            static_cast<uint32_t>(std::round((data - sec64) * 1e6))
        };
        // avoid rounding errors
        time.tv_sec += (time.tv_usec / 1000000ul);
        time.tv_usec %= 1000000ul;
        return time;
    }

    double get_bytetime ( uint32_t baudrate, const Settings& settings ) {
        return ( 1.0 + settings.bytesize + settings.parity + settings.stopbits
                + ( settings.stopbits == HALFONE ? ( 1.5 - HALFONE ) : 0.0) ) / baudrate;
    }

    speed_t get_baudrate (uint32_t baudrate) {
        switch (baudrate) {
# ifdef B0
            case 0: return B0;
# endif
# ifdef B50
            case 50: return B50;
# endif
# ifdef B75
            case 75: return B75;
# endif
# ifdef B110
            case 110: return B110;
# endif
# ifdef B134
            case 134: return B134;
# endif
# ifdef B150
            case 150: return B150;
# endif
# ifdef B200
            case 200: return B200;
# endif
# ifdef B300
            case 300: return B300;
# endif
# ifdef B600
            case 600: return B600;
# endif
# ifdef B1200
            case 1200: return B1200;
# endif
# ifdef B1800
            case 1800: return B1800;
# endif
# ifdef B2400
            case 2400: return B2400;
# endif
# ifdef B4800
            case 4800: return B4800;
# endif
# ifdef B7200
            case 7200: return B7200;
# endif
# ifdef B9600
            case 9600: return B9600;
# endif
# ifdef B14400
            case 14400: return B14400;
# endif
# ifdef B19200
            case 19200: return B19200;
# endif
# ifdef B28800
            case 28800: return B28800;
# endif
# ifdef B57600
            case 57600: return B57600;
# endif
# ifdef B76800
            case 76800: return B76800;
# endif
# ifdef B38400
            case 38400: return B38400;
# endif
# ifdef B115200
            case 115200: return B115200;
# endif
# ifdef B128000
            case 128000: return B128000;
# endif
# ifdef B153600
            case 153600: return B153600;
# endif
# ifdef B230400
            case 230400: return B230400;
# endif
# ifdef B256000
            case 256000: return B256000;
# endif
# ifdef B460800
            case 460800: return B460800;
# endif
# ifdef B500000
            case 500000: return B500000;
# endif
# ifdef B576000
            case 576000: return B576000;
# endif
# ifdef B921600
            case 921600: return B921600;
# endif
# ifdef B1000000
            case 1000000: return B1000000;
# endif
# ifdef B1152000
            case 1152000: return B1152000;
# endif
# ifdef B1500000
            case 1500000: return B1500000;
# endif
# ifdef B2000000
            case 2000000: return B2000000;
# endif
# ifdef B2500000
            case 2500000: return B2500000;
# endif
# ifdef B3000000
            case 3000000: return B3000000;
# endif
# ifdef B3500000
            case 3500000: return B3500000;
# endif
# ifdef B4000000
            case 4000000: return B4000000;
# endif
            default: return 0;
        }
    }

    void init_options ( termios * option ) {
        // INITIALIZE OPTIONS
        option->c_cflag |= (tcflag_t)  (CLOCAL | CREAD);
        option->c_lflag &= (tcflag_t) ~(ICANON | ECHO | ECHOE | ECHOK | ECHONL | ISIG | IEXTEN);
        option->c_oflag &= (tcflag_t) ~(OPOST);
        option->c_iflag &= (tcflag_t) ~(INLCR | IGNCR | ICRNL | IGNBRK);
#ifdef IUCLC
        option->c_iflag &= (tcflag_t) ~IUCLC;
#endif
#ifdef PARMRK
        option->c_iflag &= (tcflag_t) ~PARMRK;
#endif
    }

    void set_options ( termios * option, const Settings& settings ) {
        // SET CHAR LENGTH
        option->c_cflag &= (tcflag_t) ~CSIZE;
        switch ( settings.bytesize ) {
            case EIGHT: option->c_cflag |= CS8; break;
            case SEVEN: option->c_cflag |= CS7; break;
            case SIX: option->c_cflag |= CS6; break;
            case FIVE: option->c_cflag |= CS5; break;
            default: throw new std::invalid_argument ( "set_options : bytesize" );
        }
        // SET STOP BITS
        switch ( settings.stopbits ) {
            case ONE: option->c_cflag &= (tcflag_t) ~(CSTOPB); break;
            case HALFONE: option->c_cflag |=  (CSTOPB); break;
            case TWO: option->c_cflag |=  (CSTOPB); break;
            default: throw new std::invalid_argument ( "set_options : stopbits" );
        }
        // SET PARITY
        option->c_iflag &= (tcflag_t) ~(INPCK | ISTRIP);
        switch ( settings.parity ) {
            case NOPAR: option->c_cflag &= (tcflag_t) ~(PARENB | PARODD); break;
            case EVEN:
                option->c_cflag &= (tcflag_t) ~(PARODD); option->c_cflag |=  (PARENB); break;
            case ODD: option->c_cflag |=  (PARENB | PARODD); break;
#ifdef CMSPAR
            case MARK: option->c_cflag |=  (PARENB | CMSPAR | PARODD); break;
            case SPACE:
                option->c_cflag |=  (PARENB | CMSPAR); option->c_cflag &= (tcflag_t) ~(PARODD); break;
#endif
            default: throw new std::invalid_argument ( "set_options : parity" );
        }
        // SET FLOW CONTROL
        bool xonxoff=false, rtscts=false;
        switch ( settings.flowcontrol ) {
            case NOFLOW: xonxoff = false; rtscts = false; break;
            case SOFTWARE: xonxoff = true; rtscts = false; break;
            case HARDWARE: xonxoff = false; rtscts = true; break;
            default: throw new std::invalid_argument ( "set_options : flowcontrol" );
        }
        // XONXOFF
#ifdef IXANY
        if ( xonxoff ) option->c_iflag |=  (IXON | IXOFF);
        else option->c_iflag &= (tcflag_t) ~(IXON | IXOFF | IXANY);
#else
        if ( xonxoff ) option->c_iflag |=  (IXON | IXOFF);
        else option->c_iflag &= (tcflag_t) ~(IXON | IXOFF);
#endif
        // RTSCTS
#ifdef CRTSCTS
        if ( rtscts ) option->c_cflag |=  (CRTSCTS);
        else option->c_cflag &= (unsigned long) ~(CRTSCTS);
#else
        if ( rtscts ) option->c_cflag |=  (CNEW_RTSCTS);
        else option->c_cflag &= (unsigned long) ~(CNEW_RTSCTS);
#endif
        // http://www.unixwiz.net/techtips/termios-vmin-vtime.html
        // this basically sets the read call up to be a polling read,
        // but we are using select to ensure there is data available
        // to read before each call, so we should never needlessly poll
        option->c_cc[VMIN] = 0;
        option->c_cc[VTIME] = 0;
    }

    // Get file descriptor options
    int get_options ( int fd ) {
        int argument = fcntl(fd, F_GETFL, NULL);
        if ( argument < 0 )
            throw IOException ( "get file access mode", errno );
        return argument;
    }

    // Set file descriptor options
    void set_options ( int fd, int cmd ) {
        if ( fcntl(fd, F_SETFL, cmd ) < 0 )
            throw IOException ( "set", errno );
    }

    // Set socket address and port
    void set_address ( const string * address, uint32_t port, sockaddr_in * sockaddr ) {
        if ( address->empty() )
            throw invalid_argument ( "Empty ip address is invalid" );
        if ( port == 0 )
            throw invalid_argument ( "unspecified port is invalid" );
        // parse connection data into sock_addr_in struct
        sockaddr->sin_port = htons(port);
        if ( inet_pton(sockaddr->sin_family, address->c_str(), & ( sockaddr->sin_addr ) ) < 1 )
            throw InterfaceException ( "unable to parse ip address", errno );
    }

}

src/comm.cc
// this is for emacs file handling -*- mode: c++; indent-tabs-mode: nil -*-

// -- BEGIN LICENSE BLOCK -----------------------------------------------------------------------------------------------------

/*!
 *  Copyright (C) 2021, Autognity Srl, Via del Popolo 6, Orvieto (TR)
 *  \file ether.cpp
 *  \author Andrea Tamantini <andrea.tamantini@autognity.com>
 *  \date 2021-05-18
 */

// -- END LICENSE BLOCK -------------------------------------------------------------------------------------------------------


/*=============================================================================================================================
 * HEADER
 *===========================================================================================================================*/
#include <comm/comm.h>

namespace comm {

    /*! Constructor */
    Comm::Comm ( const string &address, const string& eol, Timeout timeout, Settings settings ) :
            address_(address), eol_(eol), eol_len_(eol.length()), timeout_(timeout), settings_(settings) { }
    /*! Destructor */
    Comm::~Comm () {
        boost::lock_guard<boost::mutex> lock_read(this->mtx_read);
        boost::lock_guard<boost::mutex> lock_send(this->mtx_send);
        this->close_();
    }

    /*! Opens the comm port. */
    void Comm::open () {
        std::cout << "OPEN 1" << std::endl;
        boost::lock_guard<boost::mutex> lock_read(this->mtx_read);
        boost::lock_guard<boost::mutex> lock_send(this->mtx_send);
        this->open_();
        this->connect_();
        std::cout << "OPEN 2" << std::endl;
    }
    /*! Closes the comm port. */
    void Comm::close () {
        std::cout << "CLOSE 1" << std::endl;
        boost::lock_guard<boost::mutex> lock_read(this->mtx_read);
        boost::lock_guard<boost::mutex> lock_send(this->mtx_send);
        this->close_();
        std::cout << "CLOSE 2" << std::endl;
    }

    /*! Flush the input and output buffers */
    void Comm::flush () {
        std::cout << "FLUSH 1" << std::endl;
        boost::lock_guard<boost::mutex> lock_read(this->mtx_read);
        boost::lock_guard<boost::mutex> lock_send(this->mtx_send);
        this->flush_();
        std::cout << "FLUSH 2" << std::endl;
    }
    void Comm::flushInput () {
        std::cout << "FLUSH IN 1" << std::endl;
        boost::lock_guard<boost::mutex> lock_read(this->mtx_read);
        boost::lock_guard<boost::mutex> lock_send(this->mtx_send);
        this->flushInput_();
        std::cout << "FLUSH IN 2" << std::endl;
    }
    void Comm::flushOutput () {
        std::cout << "FLUSH OUT 1" << std::endl;
        boost::lock_guard<boost::mutex> lock_read(this->mtx_read);
        boost::lock_guard<boost::mutex> lock_send(this->mtx_send);
        this->flushOutput_();
        std::cout << "FLUSH OUT 2" << std::endl;
    }

    /*! Gets the open status of the comm port.
    *
    * \return Returns true if the port is open, false otherwise.
    */
    bool Comm::isOpen () const { return this->is_open_; }
    bool Comm::isConnected () const { return this->is_connected_; }

    /*! Block until there is comm data to read or read_constant
    * number of milliseconds have elapsed. The return value is true when
    * the function exits with the port in a readable state, false otherwise
    * (due to timeout or select interruption). */
    bool Comm::waitRead () {
        boost::lock_guard<boost::mutex> lock_read(this->mtx_read);
        return this->waitRead_() > 0;
    }
    bool Comm::waitSend () {
        boost::lock_guard<boost::mutex> lock_send(this->mtx_send);
        return this->waitSend_() > 0;
    }

    /*=====================================================================================================================
     * READ, READLINE and READLINES : Public methods to read a fixed number of characters, a line or an array of lines
     *=====================================================================================================================
     * READ : Read a fixed size of char and parse them into a char array, char vector or string, returns size or string
     *-------------------------------------------------------------------------------------------------------------------*/
    // READ (char*,size) -> size : Threadsafely read a fixed size of char in a char array
    size_t Comm::read (uint8_t *buffer, size_t size) {
        std::cout << "READ UINT 1" << std::endl;
        boost::lock_guard<boost::mutex> lock(this->mtx_read);
        std::cout << "READ UINT 2" << std::endl;
        return this->read_ (buffer, size);
    }
    // READ (vector<char>,size) -> size : Threadsafely read a fixed size of char in a char vector
    size_t Comm::read (vector<uint8_t> &buffer, size_t size) {
        std::cout << "READ VEC 1" << std::endl;
        boost::lock_guard<boost::mutex> lock(this->mtx_read);
        uint8_t *buffer_ = new uint8_t[size];
        size_t bytes_read = 0;
        try { bytes_read = this->read_ (buffer_, size); }
        catch (const std::exception &e) { delete[] buffer_; throw; }
        buffer.insert (buffer.end (), buffer_, buffer_+bytes_read);
        delete[] buffer_;
        std::cout << "READ VEC 2" << std::endl;
        return bytes_read;
    }
    // READ (string,size) -> size : Threadsafely read a fixed size of char in a string
    size_t Comm::read (string &buffer, size_t size) {
        std::cout << "READ STR 1" << std::endl;
        boost::lock_guard<boost::mutex> lock(this->mtx_read);
        uint8_t *buffer_ = new uint8_t[size];
        size_t bytes_read = 0;
        try { bytes_read = this->read_ (buffer_, size); }
        catch (const std::exception &e) { delete[] buffer_; throw; }
        buffer.append (reinterpret_cast<const char*>(buffer_), bytes_read);
        delete[] buffer_;
        std::cout << "READ STR 2" << std::endl;
        return bytes_read;
    }
    // READ (size) -> string : Threadsafely read a fixed size of char in a string
    string Comm::read (size_t size) {
        std::string buffer;
        this->read (buffer, size);
        return buffer;
    }
    /*---------------------------------------------------------------------------------------------------------------------
     * READLINE : Read a fixed size of char and parse them into a char array, char vector or string, returns size or string
     *-------------------------------------------------------------------------------------------------------------------*/
    // READLINE (string,size) -> size : Read a line (until eol or size is reached) into string, return the string size
    size_t Comm::readline (string& buffer, size_t size) {
        std::cout << "READ LINE 1" << std::endl;
        boost::lock_guard<boost::mutex> lock(this->mtx_read);
        uint8_t *buffer_ = static_cast<uint8_t*> (alloca (size * sizeof (uint8_t)));
        size_t read_so_far = 0;
        while ( read_so_far < size ) {
            size_t bytes_read = this->read_ (buffer_ + read_so_far, 1);
            read_so_far += bytes_read;
            if (bytes_read == 0) break;  // Timeout occured on reading 1 byte
            if(read_so_far < this->eol_len_) continue;
            if (string (reinterpret_cast<const char*>(buffer_+read_so_far-this->eol_len_),this->eol_len_) == this->eol_)
                break; // EOL found
        }
        buffer.append(reinterpret_cast<const char*> (buffer_), read_so_far);
        std::cout << "READ LINE 2" << std::endl;
        return read_so_far;
    }
    // READLINE (size) -> string : Read a line (until eol or size is reached) into a string, return the string
    string Comm::readline ( size_t size ) {
        string buffer;
        this->readline (buffer, size);
        return buffer;
    }
    /*---------------------------------------------------------------------------------------------------------------------
     * READLINES : Read multiple lines at once, emplace them in a vector of strings
     *-------------------------------------------------------------------------------------------------------------------*/
    // READLINES (size) -> vector<string> : Read a vector of strings (to eol) until a fixed size is reached, return the string
    vector<string> Comm::readlines ( size_t size ) {
        std::cout << "READ LINES 1" << std::endl;
        boost::lock_guard<boost::mutex> lock(this->mtx_read);
        vector<string> lines;
        uint8_t* buffer_ = static_cast<uint8_t*> (alloca (size * sizeof (uint8_t)));
        size_t read_so_far = 0;
        size_t start_of_line = 0;
        while (read_so_far < size) {
            size_t bytes_read = this->read_ (buffer_+read_so_far, 1);
            if ( bytes_read == 0 ) break; // Timeout occured on reading 1 byte
            read_so_far += bytes_read;
            if ( read_so_far < this->eol_len_ + start_of_line ) continue;
            if ( strncmp ( reinterpret_cast<const char*>(buffer_+read_so_far-this->eol_len_),
                        this->eol_.c_str(), this->eol_len_ ) ) {  // EOL found
                lines.emplace_back(reinterpret_cast<const char*>(buffer_+start_of_line), read_so_far-start_of_line);
                start_of_line = read_so_far;
            }
            if ( read_so_far == size ) break; // Reached the maximum read length
        }
        if (start_of_line != read_so_far)
            lines.emplace_back(reinterpret_cast<const char*>(buffer_+start_of_line), read_so_far-start_of_line);
        std::cout << "READ LINES 2" << std::endl;
        return lines;
    }


    /*=====================================================================================================================
     * SEND : Public methods to send a frame in string, char array or char vector form
     *=====================================================================================================================
     * SEND : Send a string, char* or vector<char>. return the sent size
     *-------------------------------------------------------------------------------------------------------------------*/
    // SEND (string) -> size : Send a string, returns the number of sent char
    size_t Comm::send (const string& data) {
        std::cout << "SEND STR 1" << std::endl;
        boost::lock_guard<boost::mutex> lock(this->mtx_send);
        std::cout << "SEND STR 2" << std::endl;
        return this->send_ (reinterpret_cast<const uint8_t*>(data.c_str()), data.length());
    }
    // SEND (vector<char>) -> size : Send a char vector, returns the number of sent char
    size_t Comm::send (const std::vector<uint8_t> &data) {
        std::cout << "SEND VEC 1" << std::endl;
        boost::lock_guard<boost::mutex> lock(this->mtx_send);
        std::cout << "SEND VEC 2" << std::endl;
        return this->send_ (&data[0], data.size());
    }
    // SEND (char*,size) -> size : Send a char array, returns the number of sent char
    size_t Comm::send (const uint8_t *data, size_t size) {
        std::cout << "SEND UINT 1" << std::endl;
        boost::lock_guard<boost::mutex> lock(this->mtx_send);
        std::cout << "SEND UINT 2" << std::endl;
        return this->send_ (data, size);
    }

    /*=====================================================================================================================
     * GETTERS AND SETTERS : Public methods to set and get Comm parameters
     *===================================================================================================================*/
    // SET ADDRESS
    void Comm::setAddress (const std::string &address) {
        if ( this->address_.compare(address) == 0 ) return;
        this->address_ = address;
        boost::lock_guard<boost::mutex> lock_read(this->mtx_read);
        boost::lock_guard<boost::mutex> lock_send(this->mtx_send);
        if ( this->is_connected_ ) { this->close_(); this->open_(); }
    }
    // GET ADDRESS
    const string& Comm::getAddress () const { return this->address_; }
    //---------------------------------------------------------------------------------------------------------------------
    // SET PORT
    void Comm::setPort ( uint16_t port ) {
        if ( this->port_ == port ) return;
        this->port_ = port;
        boost::lock_guard<boost::mutex> lock_read(this->mtx_read);
        boost::lock_guard<boost::mutex> lock_send(this->mtx_send);
        if ( this->is_connected_ ) { this->close_(); this->open_(); }
    }
    // GET PORT
    uint16_t Comm::getPort () const { return this->port_; }
    //---------------------------------------------------------------------------------------------------------------------
    // SET BAUDRATE
    void Comm::setBaudrate ( uint32_t baudrate ) {
        if ( this->baudrate_ == baudrate ) return;
        this->baudrate_ = baudrate;
        boost::lock_guard<boost::mutex> lock_read(this->mtx_read);
        boost::lock_guard<boost::mutex> lock_send(this->mtx_send);
        if ( this->is_connected_ ) { this->close_(); this->open_(); }
    }
    // GET BAUDRATE
    uint32_t Comm::getBaudrate () const { return this->baudrate_; }
    //---------------------------------------------------------------------------------------------------------------------
    // SET EOL : Set end of the line char for payloads (frames) to be read
    void Comm::setEOL ( const string& eol ) {
        boost::lock_guard<boost::mutex> lock(this->mtx_read);
        this->eol_ = eol; this->eol_len_ = eol.length(); }
    // GET EOL : Get end of the line char for payloads (frames) to be read
    const string& Comm::getEOL ( ) const { return this->eol_; }
    //---------------------------------------------------------------------------------------------------------------------
    // SET TIMEOUT : Set timeout for read and send operations and connection (passing a timeout struct)
    void Comm::setTimeout (const Timeout& timeout) {
        boost::lock_guard<boost::mutex> lock_read(this->mtx_read);
        boost::lock_guard<boost::mutex> lock_send(this->mtx_send);
        this->timeout_ = timeout;
        this->setOptions_ ( );
    }
    // SET TIMEOUT : Set timeout for read and send operations and connection (specified read, send and connection timeouts)
    void Comm::setTimeout ( double read, double send, double byte, double conn ) {
        boost::lock_guard<boost::mutex> lock_read(this->mtx_read);
        boost::lock_guard<boost::mutex> lock_send(this->mtx_send);
        this->timeout_ = Timeout ( read, send, byte, conn );
        this->setOptions_ ( );
    }
    // GET TIMEOUT : Get timeout for read and send operations and connection (passing a timeout struct)
    const Timeout& Comm::getTimeout ( ) const { return this->timeout_; }
    //---------------------------------------------------------------------------------------------------------------------
    // SET SETTINGS
    void Comm::setSettings ( const Settings& settings ) {
        boost::lock_guard<boost::mutex> lock_read(this->mtx_read);
        boost::lock_guard<boost::mutex> lock_send(this->mtx_send);
        this->settings_ = settings;
        this->setOptions_();
    }
    void Comm::setSettings ( ByteSize bytesize, Parity parity, StopBits stopbits, FlowControl flowcontrol ) {
        boost::lock_guard<boost::mutex> lock_read(this->mtx_read);
        boost::lock_guard<boost::mutex> lock_send(this->mtx_send);
        this->settings_ = Settings(bytesize, parity, stopbits, flowcontrol);
        this->setOptions_();
    }
    // GET SETTINGS
    const Settings& Comm::getSettings ( ) const { return this->settings_; }

    /*=====================================================================================================================
     * VIRTUAL : Virtual private methods to be extended
     *===================================================================================================================*/
    // Read common function ( VIRTUAL )
    size_t Comm::read_ (uint8_t *data, size_t size) { return -1; }
    // Send common function ( VIRTUAL )
    size_t Comm::send_ (const uint8_t *data, size_t size) { return -1; }
    // Open file descriptor
    void Comm::open_ ( ) { throw new IOException ( "Comm::open : to be extended" ); }
    // Close connection and file descriptor ( COMMON )
    void Comm::close_ ( ) {
        if ( this->is_open_ == true) {
            if ( this->fd_ != -1) {
                int result = ::close (fd_);
                if ( result != 0 )
                    throw new IOException ( "Comm::close", errno );
                this->fd_ = -1;
            }
            this->is_connected_ = false;
            this->is_open_ = false;
        }
    }
    // Connect ( SERIAL )
    void Comm::connect_ ( ) { throw new IOException ( "Comm::connect : to be extended" ); }
    // Set Options ( SERIAL )
    void Comm::setOptions_ ( ) { throw new InterfaceException ( "Comm::setOptions : to be extended" ); }


    /*! Block until there is comm data to read or read_constant
    * number of milliseconds have elapsed. The return value is true when
    * the function exits with the port in a readable state, false otherwise
    * (due to timeout or select interruption). */
    int Comm::waitRead_ () {
        // Setup a select call to block for comm data or a timeout
        fd_set fd_set_;
        FD_ZERO ( &fd_set_ );
        FD_SET ( this->fd_, &fd_set_ );
        int result = select (fd_ + 1, &fd_set_, NULL, NULL, &(this->timeout_.conn));
        std::cout << "WAIT READ 1 : " << result << std::endl;
        if (result < 0) {
            std::cout << "WAIT READ 2 : " << errno << std::endl;
            // Select was interrupted
            if (errno == EINTR) return 0;
            // Otherwise there was some error
            throw new IOException ( "waitRead", errno );
        }
        // This shouldn't happen, if r > 0 our fd has to be in the list!
        if (result > 0 && ! FD_ISSET ( fd_, &fd_set_ ) )
            throw new IOException (
                    "Comm::waitRead : select reports ready to read, but our fd isn't in the list, this shouldn't happen!");
        // Data available to read.
        std::cout << "WAIT READ 3" << std::endl;
        return result;
    }
    int Comm::waitSend_ () {
        std::cout << "WAIT SEND 1" << std::endl;
        // Setup a select call to block for comm data or a timeout
        fd_set fd_set_;
        FD_ZERO ( &fd_set_ );
        FD_SET ( this->fd_, &fd_set_ );
        int result = select (fd_ + 1, NULL, &fd_set_, NULL, &(this->timeout_.conn));
        if (result < 0) {
            // Select was interrupted
            if (errno == EINTR) return 0;
            // Otherwise there was some error
            throw new IOException ( "waitSend", errno );
        }
        // This shouldn't happen, if r > 0 our fd has to be in the list!
        if (result > 0 && ! FD_ISSET ( fd_, &fd_set_ ) )
            throw new IOException (
                    "Comm::waitSend : select reports ready to send, but our fd isn't in the list, this shouldn't happen!");
        // Data available to read.
        std::cout << "WAIT SEND 2" << std::endl;
        return result;
    }

    // FLUSH : For the ether socket this does nothing ( SERIAL )
    void Comm::flush_ ( ) { }
    void Comm::flushInput_ ( ) { }
    void Comm::flushOutput_ ( ) { }

}